src/compiler/types.rs

```diff
@@ -314,6 +898,255 @@ pub enum ExitStatus {
+        const Intersection = 1 << 21;
+
+        const Literal = Self::StringLiteral.bits | Self::NumberLiteral.bits | Self::BigIntLiteral.bits | Self::BooleanLiteral.bits;
+        const StructuredType = Self::Object.bits | Self::Union.bits | Self::Intersection.bits;
+        const StructuredOrInstantiable = Self::StructuredType.bits /*| Self::Instantiable.bits */;
+    }
+}
+
+#[derive(Clone)]
+pub enum Type {
+    IntrinsicType(IntrinsicType),
+    UnionOrIntersectionType(UnionOrIntersectionType),
+}
+
+impl TypeInterface for Type {
+    fn flags(&self) -> TypeFlags {
+        match self {
+            Type::IntrinsicType(intrinsic_type) => intrinsic_type.flags(),
+            Type::UnionOrIntersectionType(union_or_intersection_type) => {
+                union_or_intersection_type.flags()
+            }
+        }
+    }
+}
+
+pub trait TypeInterface {
+    fn flags(&self) -> TypeFlags;
+}
+
+#[derive(Clone)]
+pub struct BaseType {
+    pub flags: TypeFlags,
+}
+
+impl TypeInterface for BaseType {
+    fn flags(&self) -> TypeFlags {
+        self.flags
+    }
+}
+
+pub trait IntrinsicTypeInterface: TypeInterface {}
+
+#[derive(Clone)]
+pub enum IntrinsicType {
+    BaseIntrinsicType(BaseIntrinsicType),
+    FreshableIntrinsicType(FreshableIntrinsicType),
+}
+
+impl TypeInterface for IntrinsicType {
+    fn flags(&self) -> TypeFlags {
+        match self {
+            IntrinsicType::BaseIntrinsicType(base_intrinsic_type) => base_intrinsic_type.flags(),
+            IntrinsicType::FreshableIntrinsicType(freshable_intrinsic_type) => {
+                freshable_intrinsic_type.flags()
+            }
+        }
+    }
+}
+
+impl IntrinsicTypeInterface for IntrinsicType {}
+
+impl From<IntrinsicType> for Type {
+    fn from(intrinsic_type: IntrinsicType) -> Self {
+        Type::IntrinsicType(intrinsic_type)
+    }
+}
+
+#[derive(Clone)]
+pub struct BaseIntrinsicType {
+    _type: BaseType,
+}
+
+impl BaseIntrinsicType {
+    pub fn new(type_: BaseType) -> Self {
+        Self { _type: type_ }
+    }
+}
+
+impl TypeInterface for BaseIntrinsicType {
+    fn flags(&self) -> TypeFlags {
+        self._type.flags()
+    }
+}
+
+impl IntrinsicTypeInterface for BaseIntrinsicType {}
+
+impl From<BaseIntrinsicType> for IntrinsicType {
+    fn from(base_intrinsic_type: BaseIntrinsicType) -> Self {
+        IntrinsicType::BaseIntrinsicType(base_intrinsic_type)
+    }
+}
+
+impl From<BaseIntrinsicType> for Type {
+    fn from(base_intrinsic_type: BaseIntrinsicType) -> Self {
+        Type::IntrinsicType(IntrinsicType::BaseIntrinsicType(base_intrinsic_type))
+    }
+}
+
+#[derive(Clone)]
+pub struct FreshableIntrinsicType {
+    _intrinsic_type: BaseIntrinsicType,
+    pub fresh_type: WeakSelf<Type>,
+    pub regular_type: WeakSelf<Type>,
+}
+
+impl FreshableIntrinsicType {
+    pub fn new(intrinsic_type: BaseIntrinsicType) -> Self {
+        Self {
+            _intrinsic_type: intrinsic_type,
+            fresh_type: WeakSelf::new(),
```

Comment:
Here's the rest of the `WeakSelf` API that I used, `WeakSelf::new()` initializes it and then below `.get()` returns the expected (already-initialized via `.init()` somewhere) weak reference (or else I guess panics if you haven't actually initialized it yet?)